### PR TITLE
Implement top-k optimization

### PR DESCRIPTION
**Pull request by acquamarin**
This PR implements the optimization for top-k queries.
Sample top-k queries:
`MATCH (comment:Comment) return comment.length,comment.creationDate ORDER BY comment.length, comment.creationDate LIMIT 5`
1. Plan without top-k optimizatioin: We firstly accumulate and order by all tuples, then we perform a limit operation on the result.
2. Plan with top-k optimization:
We merge the order by and limit operator together.
Instead of accumulating all tuples, we do local sort and only keep the top-k tuples in each thread.

Performance number:
https://docs.google.com/spreadsheets/d/1K53Yz8KMuvrFfXbyoPyULhWt9nSBEUI6W4WdMWirODM/edit#gid=1649459535

### CODE REVIEW

File: src/processor/operator/order_by/key_block_merger.cpp [hunk 1]
```diff
@@ -45,6 +45,8 @@ void BlockPtrInfo::updateTuplePtrIfNecessary() {
             curTuplePtr = keyBlocks->getKeyBlockBuffer(curBlockIdx);
             curBlockEndTuplePtr =
                 keyBlocks->getBlockEndTuplePtr(curBlockIdx, endTupleIdx, endBlockIdx);
+        } else if (curBlockIdx > endBlockIdx) {
```
Comment (acquamarin) [inline src/processor/operator/order_by/key_block_merger.cpp:48]: remove

File: src/include/processor/operator/order_by/top_k.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#include "processor/operator/sink.h"
+#include "sort_state.h"
+
+namespace kuzu {
+namespace processor {
+
+using vector_select_comparison_func =
+    std::function<bool(common::ValueVector&, common::ValueVector&, common::SelectionVector&)>;
+
+struct TopKScanState {
```
Comment (andyfengHKU) [inline src/include/processor/operator/order_by/top_k.h:15]: Leave a TODO for me to move this into mapper

**Review by andyfengHKU — APPROVED** [review]
Let's fix the coverage for this PR before merging

File: src/include/processor/operator/order_by/order_by_key_encoder.h [hunk 3]
```diff
@@ -28,6 +28,29 @@ namespace processor {
...
+        : OrderByDataInfo{other.keysPosAndType, other.payloadsPosAndType, other.isPayloadFlat,
+              other.isAscOrder, other.mayContainUnflatKey} {}
+
+public:
+    std::vector<std::pair<DataPos, common::LogicalType>> keysPosAndType;
+    std::vector<std::pair<DataPos, common::LogicalType>> payloadsPosAndType;
+    std::vector<bool> isPayloadFlat;
+    std::vector<bool> isAscOrder;
+    // TODO(Ziyi): We should figure out unflat keys in a more general way.
+    bool mayContainUnflatKey;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/processor/operator/order_by/order_by_key_encoder.h:51]: #1965 opened

File: src/include/processor/operator/order_by/sort_state.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+#include "processor/operator/order_by/radix_sort.h"
+#include "processor/result/factorized_table.h"
+
+namespace kuzu {
+namespace processor {
+
+struct LocalPayloadTableInfo {
+    uint64_t globalIdx;
+    FactorizedTable* payload;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/processor/operator/order_by/sort_state.h:13]: The globalIdx always comes with payload, and the struct is returned by 
```
 LocalPayloadTableInfo getLocalPayloadTable(storage::MemoryManager& memoryManager);
```
Better to keep them together inside a struct.

File: src/include/processor/operator/order_by/top_k.h (hunk 2 above)
Comment (acquamarin) [reply src/include/processor/operator/order_by/top_k.h:15]: I don't think we can move the initialization to mapper.
We need have access to the sortedKeyBlock. If the sorting hasn't started, we don't have access to the sortedKeyBlock.

**Pull request merged**